src/storage/index/IndexExecutor.inl

```diff
@@ -0,0 +1,329 @@
+}
+
+template <typename RESP>
+cpp2::ErrorCode IndexExecutor<RESP>::checkReturnColumns(const std::vector<std::string> &cols) {
+    int32_t index = 0;
+    if (!cols.empty()) {
+        schema_ = std::make_shared<SchemaWriter>();
+        auto schema = isEdgeIndex_ ?
+                      schemaMan_->getEdgeSchema(spaceId_, tagOrEdge_) :
+                      schemaMan_->getTagSchema(spaceId_, tagOrEdge_);
+        if (!schema) {
+            VLOG(3) << "Can't find schema, spaceId "
+                    << spaceId_ << ", id " << tagOrEdge_;
+            return isEdgeIndex_ ? cpp2::ErrorCode::E_TAG_PROP_NOT_FOUND :
+                                  cpp2::ErrorCode::E_EDGE_PROP_NOT_FOUND;
+        }
+
+        for (auto i = 0; i < static_cast<int64_t>(schema_->getNumFields()); i++) {
+            PropContext prop;
+            const auto& name = schema_->getFieldName(i);
+            auto ftype = schema_->getFieldType(name);
+            prop.type_ = ftype;
+            prop.retIndex_ = index++;
+            storage::cpp2::PropDef pd;
+            pd.set_name(name);
+            prop.prop_ = std::move(pd);
+            prop.returned_ = true;
+            props_.emplace_back(std::move(prop));
+        }
+        for (auto& col : cols) {
+            const auto& ftype = schema->getFieldType(col);
+            if (UNLIKELY(ftype == CommonConstants::kInvalidValueType())) {
+                return cpp2::ErrorCode::E_IMPROPER_DATA_TYPE;
+            }
+            /**
+             * Create PropContext for getRowFromReader
+             */
+            {
+                PropContext prop;
+                prop.type_ = ftype;
+                prop.retIndex_ = index++;
+                storage::cpp2::PropDef pd;
+                pd.set_name(col);
+                prop.prop_ = std::move(pd);
+                prop.returned_ = true;
+                props_.emplace_back(std::move(prop));
+            }
+            schema_->appendCol(col, std::move(ftype).get_type());
+        }   // end for
+    }
+    return cpp2::ErrorCode::SUCCEEDED;
+}
+
+template <typename RESP>
+kvstore::ResultCode IndexExecutor<RESP>::executeExecutionPlan(PartitionID part) {
+    std::string prefix = NebulaKeyUtils::indexPrefix(part, index_->get_index_id())
+                        .append(prefix_);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = this->kvstore_->prefix(spaceId_,
+                                      part,
+                                      prefix,
+                                      &iter);
+    if (ret != nebula::kvstore::SUCCEEDED) {
+        return ret;
+    }
+    while (iter->valid() &&
+           rowNum_ < FLAGS_max_rows_returned_per_lookup) {
+        auto key = iter->key();
+        /**
+         * Need to filter result with expression if is not accurate scan.
+         */
+        if (isIndexScan_ && !conditionsCheck(key)) {
+            iter->next();
+            continue;
+        }
+        ret = getDataRow(part, key);
```

Comment:
We'd better find all valid edges/vertices firstly,  then get the properties from then.